xapian-letor/scorer/err_score.cc

```diff
@@ -0,0 +1,94 @@
+/** @file err_score.cc
+ *  @brief Implementation of ERRScore
+ */
+/* Copyright (C) 2014 Hanxiao Sun
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian-letor/scorer.h"
+
+#include "debuglog.h"
+
+#include <algorithm>
+
+using namespace std;
+
+using namespace Xapian;
+
+ERRScore::ERRScore()
+{
+    LOGCALL_CTOR(API, "ERRScore", NO_ARGS);
+}
+
+ERRScore::~ERRScore()
+{
+    LOGCALL_DTOR(API, "ERRScore");
+}
+
+static vector<double>
+get_labels(vector<Xapian::FeatureVector> fvv)
+{
+    LOGCALL_STATIC(API, vector<double>, "get_labels", fvv);
+    size_t fvvsize = fvv.size();
+    vector<double> labels;
+
+    for (size_t i = 0; i < fvvsize; ++i) {
+	labels.push_back(fvv[i].get_label());
+    }
+
+    return labels;
+}
+
+/* test the err_Score() use the data from goo.gl/LxEmPZ
+ * input:[3,2,4]
+ * output:0.63
+ */
+double
+ERRScore::score(const std::vector<FeatureVector> & fvv) const
+{
+    LOGCALL(API, double, "ERRScore::score", fvv);
+
+    // hard code for the a five-point scale, the 16 means 2^(5-1)
```

Comment:
It takes some effort to understand what's going on here, and the comment doesn't help. Digging into the original paper, I'm unconvinced that we've implemented anything particularly similar to ERR, so at the least some work is required to make it clear that we have. It'd be easier to follow if our code followed the naming convention of the paper as far as possible.

While we're here, the paper should be referenced in the header of the file, and maybe in the docstring for the class.

Reply:
I am not eliminating this comment for now as it does describe the magic number MAX_LABEL from the paper

Reply:
Also, in C++ `2^(5-1)` is `6` not `16` (since `^` is bit-wise exclusive or).  If you mean "raise to the power of", it'd be much better to write something that actually means that in C++, for example: `pow(2, 5 - 1)`